### ISSUE TITLE
Setter on Transaction.Current doesn't respect TransactionScopeAsyncFlowOption

### ISSUE THREAD

**Issue by thomasdc** [dotnet/corefx#37402]
When reassigning `Transaction.Current`, I seem to lose the original `TransactionScopeAsyncFlowOption` behavior of my `TransactionScope`. The code after the `await` loses it's `Transaction.Current` value.

Sample code:

```
async Task Main()
{
	var scope = new TransactionScope(TransactionScopeOption.Required, TransactionScopeAsyncFlowOption.Enabled);
	var transaction = Transaction.Current;

	Transaction.Current = transaction;
	Debug.Assert(Transaction.Current != null);

	await Task.Delay(10);

	Debug.Assert(Transaction.Current == null); // Expected Transaction.Current to be NOT null since transaction.TransactionScopeAsyncFlowOption == TransactionScopeAsyncFlowOption.Enabled
}
```

I [know](https://docs.microsoft.com/en-us/dotnet/api/system.transactions.transaction.current?view=netframework-4.8) I should be using a `using` statement on `TransactionScope` rather than reassigning  the ambient transaction, however because of reasons, I'm not possible to do that. I'm curious about the reason of the behavior of the snippet above and wonder if there are ways to keep the original `TransactionScopeAsyncFlowOption` behavior.

**Comment by jimcarley** [comment]
The TransactionScopeAsyncFlowOption applies to the TransactionScope. Manually modifying Transaction.Current while there is an "active" TransactionScope resets the "logical call context" data and reverts back to assuming TLS usage.

Could you elaborate on the reasons why you need to manually set Transaction.Current while using TransactionScope?

Remember that there is a TransactionScope constructor which accepts a Transaction object that is set to Transaction.Current by the TransactionScope constructor. And if you want to make Transaction.Current null while you do some operations, you can create a "nested" TransactionScope with TransactionScopeOption.Suppress. When this "suppressing" scope is disposed, then the transaction from the "parent" TransactionScope would be put back into Transaction.Current.

**Comment by thomasdc** [comment]
This [issue](https://github.com/rebus-org/Rebus.TransactionScopes/issues/7) occurred in an open-source messaging framework called [Rebus](https://github.com/rebus-org/Rebus).

I was using a relational database table as a queue and wanted the message handling and (de)queueing to happen in the same ambient transaction.

However, the frameworks polling for new messages and handling happens on seperate threads. The transaction was transferred from the polling thread to the handler thread and restored as an ambient transaction via the `Transaction.Current` setter causing it to apparently lose its state across async continuations. Even @StephenCleary was suprised by this behavior (see [this](https://stackoverflow.com/questions/55955218/context-is-lost-when-reassigning-transaction-current-in-async-await-code) post on Stackoverflow).

I've solved the issue by using a `TransactionScope` constructor accepting a `Transaction`, however, I'm wondering why `Transaction.Current` has a public setter when it exhibits such an unstable behavior when used in combination with async/await.

**Comment by jimcarley** [comment]
I suspect this all stems from the need for backward compatibility in "full" .NET when TransactionScopeAsyncFlowOption was introduced.

As you know, before that time, Transaction.Current was kept track of in Thread Local Storage (TLS). So setting of Transaction.Current modified the TLS entry. But this obviously presented problems with Tasks. So the TransactionScopeAsyncFlowOption was introduced when using TransactionScope. But if a program chose to use the Transaction.Current setter behavior, we reverted back to the old behavior of TLS usage and the logical call context data is cleared. Mixing TransactionScopeAsyncFlowOption and manually setting Transaction.Current don't "mix".

If the Transaction.Current setter behavior was modified to deal with logical call context also, then existing applications might have issues with the change in behavior. And there was no way to "opt-in" to the new behavior with the Transaction.Current setter. I am not saying it is right, but rather we needed to retain compatibility for applications that were managing Transaction.Current themselves, without TransactionScope. By making TransactionScopeAsyncFlowOption an "opt-in" feature for TransactionScope, existing applications that didn't use the new feature were unaffected.

Bottom line - the Transaction.Current setter retained its existing behavior of using TLS when TransactionScopeAsyncFlowOption was introduced. In general, mixing TransactionScope and Transaction.Current setter is not a good idea and can also be problematic if Transation.Current is set to a value that does not match what the TransactionScope expects when the TransactionScope is disposed.

**Comment by thomasdc** [comment]
OK, thanks for the clarification! I guess this issue can be closed then.
